Theoretical records

ItIsJordan opened the pull request:
This PR closes #58 and closes #323.

This command executes the data migration SQL found in the [`scripts/hepdata_related_update.sql`](https://github.com/HEPData/hepdata/blob/theoretical-records/scripts/hepdata_related_update.sql) file:
```bash
psql -U hepdata -h localhost -d hepdata -f scripts/hepdata_related_update.sql
```

ItIsJordan committed:
Add related tables to submission model

Adds the RelatedTable class to the submission model containing entries for related table DOIs.

ItIsJordan committed:
Add new related tables field to the view

Adds the new related_tables variable to the record view.

ItIsJordan committed:
Update data submission process (submission.py)

Updates the process_data_file function to handle the new related_table DOI entry for new submission objects.

ItIsJordan committed:
Update record display html

Updates table_details.html to display related doi list div.

ItIsJordan committed:
Merge branch 'master' into theoretical-records

ItIsJordan committed:
Update table contents data

The previous related_tables code was setting its value in the table_contents data to be a collection of RelatedTable objects. It is now a list of strings containing doi entries.

ItIsJordan committed:
Add related_tables data to record output

Adds the data from related_tables into the generate_table_structure function to be properly returned.

ItIsJordan committed:
Update hepdata_tables.js

Updates ajax used to render HEPData tables to include related table doi rendering.

ItIsJordan committed:
Update table_details.html

I misunderstood that the data would be accessed from the context. Updated HTML to reflect this.

ItIsJordan committed:
Merge branch 'main' into theoretical-records

ItIsJordan committed:
Add related recid to submission models

Adds the related recid functionality to the models.py file in submission. Modifies the HEPSubmission class.

ItIsJordan committed:
Update records api.py

Updates records api to add related recids variable to the context for rendering.

ItIsJordan committed:
Update submission.py

Add functionality to move related recid data from the collected submission info document variable into the database.

ItIsJordan committed:
Update publication_record.html

Includes html tags for recids relation. Adds related recid section to the left side panel in the record page. Also adds placeholder for second section.

ItIsJordan committed:
Merge branch 'main' into theoretical-records

ItIsJordan committed:
Add get_related_to_this function to HEPSubmission class.

Adds a function to retrieve all records in the RelatedRecId table where the related object recid is the same.

ItIsJordan committed:
Add frontend functionality for related_to_this records

Adds functionality for recid relation. Adds support for js/html rendering of other relating records.

ItIsJordan committed:
Update related recid variable

Updates code to ensure the value saved under RelatedRecId->this_recid to be the recid as an integer. Changed from using a doi string.

ItIsJordan committed:
Update model (RelatedRecId)

Updates the RelatedRecId class in the database model to use Integer instead of String(128). This enforces reference by HEPData ID, and not inspire ID.

ItIsJordan committed:
Hide Related To elements when empty

Adds functionality hide the related recids, and related table doi object elements when empty.

ItIsJordan committed:
Update records found text.

Updates the record found text to be more accurate.

ItIsJordan committed:
Add "related to this" for DataSubmission DOI

Implements the query to get other DataSubmission objects relating to the self object. Fixed placeholder empty list for this function too.

ItIsJordan committed:
Fix sandbox related record filtering

Fixed an issue where the previous sandbox filtering logic was incorrect. Now fixed and updated comments.

ItIsJordan committed:
Fix broken test in records_test.py

Fix now broken test_data_processing test as new related data lists were not set.

ItIsJordan committed:
Rename and fix `get_related` functions

Renames the functions in HEPSubmission and DataSubmission used for retrieving other related HS/DS objects. Both functions now properly filter by checking the relevant HEPSubmission object's overall_status value. Both functions also now return the objects themselves.

ItIsJordan committed:
Add related recid/doi test function.

Adds a function to test the related record ID and table DOI submission and linking functionality.

Adds 3 test submissions based on test_submission, but with entries for related record ID and table DOIs in Submission.yaml

ItIsJordan committed:
Update hepdata-validator package to use git

Updates the hepdata-validator package in requirements.txt to use the new git branch.

ItIsJordan committed:
Related test update

Updates the test_related_records function and its test data.

- Removes useless data in the testing data, now down to a skeleton submission.

- Moves tests to a new directory for cleanliness, add 4th test to check for invalid data DOI.

ItIsJordan committed:
Remove blank line

Removes a blank line from views.py that is not needed.

ItIsJordan committed:
Add e2e test for related data (Recid) testing

Adds e2e test to test record page display of related record ID and data DOI values. This commit has completed record ID testing.

ItIsJordan committed:
Fix docstring

Fixes the docstring causing doc builds to fail.

ItIsJordan committed:
Remove some accidental code

Removes some lines accidentally left inside of the test.

ItIsJordan committed:
Merge branch 'main' into theoretical-records

ItIsJordan committed:
Remove related data count

Removes the text displaying the number of related objects found on the records page.

ItIsJordan committed:
Update e2e test for related data

Updates the related data e2e test to now test related DOI object display

ItIsJordan committed:
Adds a function to cleanup related record ID entries

Adds a function to cleanup related record ID entries as related record ID data was not removed upon an upload error, or the submission of new data.

ItIsJordan committed:
Fix related record reupload bug

Removes these redundant lines which I believe were causing an error when reuploading submissions with a related record id/data doi entry.

The DataResources of the HEPSubmission object are cleaned up in another function also.

ItIsJordan committed:
Fix html text for related records

Adds in accidentally removed text to denote the related record section of the record page.

ItIsJordan committed:
Merge branch 'main' into theoretical-records

ItIsJordan committed:
Merge branch 'main' into theoretical-records

ItIsJordan committed:
Merge branch 'main' into theoretical-records

ItIsJordan committed:
Add comment in cleanup_submission

Adds a comment explaining that related record entries are cleared in cleanup_submission and why.

ItIsJordan committed:
Fix data resource cleanup in process_general_submission_info

Fixes a bug created in f59a60f3 where removal of data resource lines caused additional resources to only be deleted if new additional resource data is deleted. This cleanup also needs to happen outside of the later if statement.

ItIsJordan committed:
Add related table update SQL

Add the SQL required to update an existing database to the updated schema version.

ItIsJordan committed:
Improve html tags for related data

Improves html tags and updates tests to reflect changes.

ItIsJordan committed:
Add scrolllable too related lists

Adds overflow to html the related list.

ItIsJordan committed:
Add removed break tag

Adds a removed break tag back into publication_record.html

ItIsJordan committed:
Remove spaces in yaml files

Removes spaces on array entries in the related test yaml files. Also removes a number of blank tables where "---" was at the end of some testing files.

ItIsJordan committed:
Update related record/iD text.

Improves clarity of the titles associated with related recid/doi items on the records page html.

ItIsJordan committed:
Merge branch 'main' into theoretical-records

ItIsJordan committed:
Update record page reference header text

Improves the header text language and styling of the record page related recid/doi sections.

ItIsJordan committed:
Update records page html

Update the records page, improving the text size and language on related data titles.

sjmf commented on hepdata/modules/submission/models.py:
```diff
@@ -117,6 +117,25 @@ class HEPSubmission(db.Model):
+
+    def get_related_hepsubmissions(self):
+        """
+        Queries the database for all records in the RelatedRecId table
```
Good use of commenting! :)

sjmf commented on hepdata/modules/submission/models.py:
```diff
@@ -117,6 +117,25 @@ class HEPSubmission(db.Model):
+        return related_submissions
+
 
 datafile_identifier = db.Table(
```
TODO: Add a comment explaining to the next developer what these do

sjmf commented on hepdata/modules/submission/models.py:
```diff
@@ -161,6 +194,9 @@ class DataSubmission(db.Model):
 
     doi = db.Column(db.String(128), nullable=True)
 
+    related_tables = db.relationship("RelatedTable", secondary="relatedtable_identifier",
```
To comment

sjmf commented on scripts/hepdata_related_update.sql:
Autogenerated

sjmf commented on tests/e2e/test_records.py:
```diff
@@ -239,6 +248,87 @@ def test_record_update(live_server, logged_in_browser):
+        recid/doi links display on the records page and link correctly.
+    """
+    browser = logged_in_browser
+    # Dictionary to store the generated data.
```
Great! :)

sjmf commented on tests/e2e/test_records.py:
```diff
@@ -239,6 +248,87 @@ def test_record_update(live_server, logged_in_browser):
+        # Create a mock DataSubmission object
+        doi_string = f"{HEPDATA_DOI_PREFIX}/hepdata.{test['recid']}.v1/t1"
+        datasubmission = DataSubmission(
+            name="Test", location_in_publication="Somewhere", data_file=1,
```
Each param on a new line?

sjmf commented on tests/e2e/test_records.py:
```diff
@@ -239,6 +248,87 @@ def test_record_update(live_server, logged_in_browser):
+            name="Test", location_in_publication="Somewhere", data_file=1,
+            publication_recid=test['recid'], associated_recid=test['recid'],
+            doi=doi_string, version=1, description="Test")
+        # Create the mock RelatedTable object and set the doi
```
Maybe a new line here too

sjmf commented on tests/e2e/test_records.py:
```diff
@@ -239,6 +248,87 @@ def test_record_update(live_server, logged_in_browser):
+        # Create the mock RelatedTable object and set the doi
+        related_doi_string = f"{HEPDATA_DOI_PREFIX}/hepdata.{test['related_recid']}.v1/t1"
+        related_table = RelatedTable(table_doi=doi_string, related_doi=related_doi_string)
+        datasubmission.related_tables.append(related_table)
```
Comment to say it's creating a related table

sjmf commented on tests/e2e/test_records.py:
```diff
@@ -239,6 +248,87 @@ def test_record_update(live_server, logged_in_browser):
+            html_element = browser.find_element(By.ID, element['element'])
+            data_list = html_element.find_element(By.CLASS_NAME, 'related-list')
+            list_items = data_list.find_elements(By.TAG_NAME, 'li')
+            assert len(list_items) == 1
```
Comment to explain assertions

sjmf commented on tests/submission_test.py:
Brilliant 👍

sjmf commented on tests/test_data/related_submission_test/related_submission_1/submission.yaml:
Is the YAML indentation in this file correct? I'd generally indent arrays by two spaces per indent level.

sjmf commented on requirements.txt:
Is the hepdata-validator git link in requirements temporary? @GraemeWatt

GraemeWatt replied:
Thanks for your code review, @sjmf!  Yes, installing `hepdata-validator` from a branch is temporary.  I'll merge HEPData/hepdata-validator#52 and make a new release 0.3.4, then we can update `requirements.txt` before this PR is merged.

sjmf commented on hepdata/modules/theme/assets/scss/record.scss:
```diff
@@ -657,6 +657,20 @@
   }
 }
 
+.related-list-container {
```
This is a really cool way of achieving left-hand scrollbar functionality.

sjmf commented on hepdata/modules/records/views.py:
```diff
@@ -297,13 +298,10 @@ def get_table_details(recid, data_recid, version):
-
-    datasub_query = DataSubmission.query.filter_by(id=data_recid,
-                                                   version=version)
+    datasub_query = DataSubmission.query.options(joinedload('related_tables')).filter_by(id=data_recid, version=version)
```
Comment about joinedload allowing query of data in other table without a second database access

sjmf commented on hepdata/modules/records/assets/js/hepdata_tables.js:
```diff
@@ -90,9 +90,10 @@ HEPDATA.table_renderer = {
-
         $("#table_description").html((table_data.description.indexOf('.') == 0) ? '' : table_data.description.trim());
 
+        HEPDATA.table_renderer.render_related_dois(table_data.related_tables, "#related-tables");
```
Comment?

sjmf commented on hepdata/modules/records/utils/submission.py:
```diff
@@ -426,6 +453,8 @@ def process_submission_directory(basepath, submission_file_path, recid,
         # Side effect that reviews will be deleted between uploads.
         cleanup_submission(recid, hepsubmission.version, added_file_names)
 
+        # Counter to store current table number.
```
Add a comment: 
This is required because the first table is the main submission. Subsequent tables passed to `process_data_file` need to be appended to the DOI, so we need to keep track of the table number!

sjmf reviewed (changes requested):
Some small bits and bobs to do, but looks 99.99% good to me!

ItIsJordan committed:
Update commenting

Updates commenting in response to PR feedback.

ItIsJordan committed:
Merge branch 'main' into theoretical-records

GraemeWatt commented on hepdata/modules/records/templates/hepdata_records/publication_record.html:
```diff
@@ -108,14 +108,37 @@
                                target="_blank">https://doi.org/{{ hepdata_doi }}</a>
                         </div>
 
-                        <br/>
```
This `<br/>` was the line space I asked you to reinsert last week, which you did in ca20d84ebe03f1d5270e1dfa04246e5d73dc64c6 but then removed again in 49015432658803135cb51f7959eed010afec26b3.  I think the line space is needed in the normal case where there are *no* related records, although it looks OK without the line space if there *are* related records.  So maybe include the line break conditionally like:
```jinja2
{% if not (ctx.related_recids or ctx.related_to_this_recids) %}
<br/>
{% endif %}
```

GraemeWatt commented on hepdata/modules/records/templates/hepdata_records/publication_record.html:
```diff
@@ -108,14 +108,37 @@
+                                </ul>
+                            </div>
+                        </div>
+                        {% endif %}
```
There's another template in the same directory called `related_record.html` that is accessed when clicking on individual tables in search results or when clicking on individual table DOIs.  Could you please add the related record information on that page too?  To avoid duplication, you could put the information in a file `components/related_recids.html` and `include` it in a similar way to `components/journal_info.html`.

GraemeWatt commented on hepdata/modules/records/utils/submission.py:
```diff
@@ -292,15 +315,19 @@ def process_general_submission_info(basepath, submission_info_document, recid):
 
+    if hepsubmission.overall_status not in ("sandbox", "sandbox_processing"):
+        if 'related_to_hepdata_recids' in submission_info_document:
+            for related_id in submission_info_document['related_to_hepdata_recids']:
```
Now change `related_to_hepdata_recids` to `related_to_hepdata_records` to match latest `hepdata-validator` release v0.3.4.  Other variables like the database fields do not need to be changed.

GraemeWatt commented on hepdata/modules/submission/models.py:
```diff
@@ -169,6 +207,44 @@ class DataSubmission(db.Model):
+
+class RelatedRecid(db.Model):
+    """
+    The submission object associated with a related table entry.
```
The comment here is copied from `RelatedTable`.  Change it for `RelatedRecid`.

GraemeWatt commented on requirements.txt:
```diff
@@ -3,7 +3,7 @@ bleach==6.0.0
 gunicorn==21.2.0
 hepdata-converter-ws-client==0.2.2
-hepdata-validator==0.3.3
+hepdata-validator @ git+https://github.com/HEPData/hepdata-validator@submission-linking
```
You can update this to `hepdata-validator==0.3.4` now.  I've checked that, after the upgrade, I can now use [`import-records`](https://hepdata.readthedocs.io/en/latest/cli.html#hepdata-importer-import-records) to import records with large tables given as examples in #136 into my local instance.

GraemeWatt commented on tests/test_data/related_submission_test/related_submission_1/submission.yaml:
```diff
@@ -0,0 +1,30 @@
+---
+description: "Test Data"
+comment: Test
+related_to_hepdata_recids:
```
Change to `related_to_hepdata_records` now.

GraemeWatt commented on tests/test_data/related_submission_test/related_submission_2/submission.yaml:
```diff
@@ -0,0 +1,30 @@
+---
+description: "Test Data"
+comment: Test
+related_to_hepdata_recids:
```
Change to `related_to_hepdata_records` now.

GraemeWatt commented on tests/test_data/related_submission_test/related_submission_3/submission.yaml:
```diff
@@ -0,0 +1,14 @@
+---
+description: "Test Data"
+comment: Test
+related_to_hepdata_recids:
```
Change to `related_to_hepdata_records` now.

GraemeWatt reviewed (changes requested):
This looks great, @ItIsJordan!  Thanks for all your hard work on this.  I made a few comments that are hopefully fairly easy to address.  Please also check Samantha's comments from last week: there were a few that I've left unresolved as they don't seem to be addressed in 6af6ee2bd41dee42d89fe2e5877862a06821eb78 (you might need to log in to GitHub to see all the comments).  Please also merge the latest updates from the `main` branch.

ItIsJordan committed:
Revert hepdata-validator git requirement & update

Update the hepdata-validator package in requirements.txt to use the updated version (0.3.4), and not the git branch.

ItIsJordan committed:
Update schema naming for related records

Updates code where accessing submission document as the schema has now renamed "related_to_hepdata_recids" to "related_to_hepdata_records".

ItIsJordan committed:
Update commenting

Updates commenting/code readability for related data testing, models and JS.

ItIsJordan committed:
Update related record html

Updates the related record html to now be contained within a component template (related_recids.html). Updates publication_record.html to use this. Also updates related_record.html to include related record data.

Adds a conditional break tag to fix formatting when related record data does not exist.

GraemeWatt committed:
records: expand ctx for related_record.html

* Need to add related_recids and related_to_this_recids to ctx.

GraemeWatt reviewed (approved):
Thanks for addressing my comments, @ItIsJordan.  In testing I found that pages displayed using the `related_record.html` template were missing expected links to related records.  This was because the `ctx` is set differently from `publication_record.html`, so the `related_recids` and `related_to_this_recids` variables need to be added separately.  I've pushed a fix in commit 4afa0949242e2c910c82abde30392de14bcae335.  I'll merge now and deploy next week.
